### ISSUE TITLE
How to improve the model for custom data

### ISSUE THREAD

**Issue by zpmmehrdad** [WongKinYiu/CrossStagePartialNetworks#21]
@WongKinYiu Hi,

I have a problem that sometimes some pictures are not detected or detected wrong. I attached my model and some images for testing. Could you please check it and guide me? I have about 2K images per class. Please give me some information about the hyperparameters for my case.

[file](https://drive.google.com/open?id=1TTElKrcm_6BZ8ty6QQDUGJuD6xnlrKFS)

Thanks  in advance

_Originally posted by @zpmmehrdad in https://github.com/WongKinYiu/CrossStagePartialNetworks/issues/6#issuecomment-616100822_

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

Would you like me to give you my dataset?

Thanks

**Comment by WongKinYiu** [comment]
hello, 

i do not see the information about counters_per_class.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

You can remove "counters_per_class" from the file. @AlexeyAB Added this feature for balancing the dataset but I don't use that for this case. I add the repeated file to the train.txt for balancing.

**Comment by WongKinYiu** [comment]
oh, i just need the `counter_per_class` information for analyzing the reason.

could you calculate the anchors with letter box?

**Comment by zpmmehrdad** [comment]
> oh, i just need the `counter_per_class` information for analyzing the reason.

I haven't trained the model with  `counter_per_class`. You can remove the line.
> could you calculate the anchors with letter box?
Yes, I can

Can I give you my small dataset?

**Comment by zpmmehrdad** [comment]
@WongKinYiu 

The detection result is good but the classified result is not good.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

Could you please give your email address to share the dataset?

Thanks in advance

**Comment by WongKinYiu** [comment]
![image](https://user-images.githubusercontent.com/12152972/80066412-7a1f7480-856e-11ea-98c9-1028f17444c4.png)

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

I sent you an invitation link. I have a lot of classes that are the same and I want to select the best model and hyperparameters for this case. Thank you for helping

Thanks in advance

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

Did you see the dataset, How can I improve it?

Thanks

**Comment by WongKinYiu** [comment]
@zpmmehrdad 

sorry for late reply.
yes, i downloaded the dataset.
i will take a look tonight.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

Thanks for the reply, I just wanted to say that I have a lot of sign which are the same. I opened an issue [Link](https://github.com/AlexeyAB/darknet/issues/5086) for example.

**Comment by WongKinYiu** [comment]
for traffic sign or licence plate or printed text, usually we use data synthesis with perspective effect for data augmentation.

but i think another problem of your dataset is height-width-ratio and small object, i will check it.

**Comment by WongKinYiu** [comment]
i think your objects are too small and too sparse.
and the dataset is consist of several groups of fine-grained classes
so
1. anchors assignment should be modified
2. maybe crop the image to make objects not so sparse
3. maybe you can consider https://arxiv.org/abs/1902.07296 and data synthesis
4. make it become multi-label problem

**Comment by zpmmehrdad** [comment]
I wanted to have different angles and backgrounds in the dataset. Does it need to resize the images before training?

> i think your objects are too small and too sparse.
> and the dataset is consist of several groups of fine-grained classes
> so
> 
> 1. anchors assignment should be modified

How can I do that?
> 2. maybe crop the image to make objects not so sparse

Should I do crop all of the images or add cropped images to the dataset?
> 3. maybe you can consider https://arxiv.org/abs/1902.07296 and data synthesis

> 4. make it become multi-label problem

Can you explain about it?

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

I always use this command for anchor :
darknet detector calc_anchors a.obj -num_of_clusters 9 -width 512 -height 512

Could you please more suggestions or give me the best CFG  with hyperparameters?

Thanks

**Comment by WongKinYiu** [comment]
[trythisfirst.txt](https://github.com/WongKinYiu/CrossStagePartialNetworks/files/4535440/trythisfirst.txt)

**Comment by zpmmehrdad** [comment]
> [trythisfirst.txt](https://github.com/WongKinYiu/CrossStagePartialNetworks/files/4535440/trythisfirst.txt)

@WongKinYiu  Thanks for the reply,
What pre-trained should I use?

**Comment by WongKinYiu** [comment]
I do not change your backbone, so just use the command you used before for training.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

Could you please change this file because 319 classes take too long for training, I want to process on 6 classes first

[v4.txt](https://github.com/WongKinYiu/CrossStagePartialNetworks/files/4535564/v4.txt)
[nest_panet_opt.txt](https://github.com/WongKinYiu/CrossStagePartialNetworks/files/4535565/nest_panet_opt.txt)

.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

I trained a model on 6 classes that I shared the dataset to you but the result is not good. Classify is not a good result.

**Comment by WongKinYiu** [comment]
could you show the map results? thanks.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

I sent you an email. the mAP is on the training dataset, not test.

**Comment by WongKinYiu** [comment]
@zpmmehrdad Hello,

Please do not set batchsize=1 and subdivision=1, please use batchsize=64 and subdivision=64 instead. You can divide max_batches and steps by 64 to train same number of images.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

Yes, I'm using batch=64 and subdivision=32 for training and use 1 for testing.

**Comment by WongKinYiu** [comment]
![image](https://user-images.githubusercontent.com/12152972/80351939-b6700f00-88a5-11ea-9ded-38a75df3bbcc.png)
this case is really strange, do the wrong classify results are all 50 km pees limit?

**Comment by zpmmehrdad** [comment]
No, Sometimes some signs are detected wrong that not in the dataset.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

What do you think I remove the poor quality images from the dataset? because when the object is far, the result usually is wrong but when the object is near the result is good, of course not always

**Comment by WongKinYiu** [comment]
yes you can do that, but if there is an automatic method can enhance performance is better.
i think in your case softmax with cross entropy loss is better than binary cross entropy loss.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

Could you please modify the CFG with softmax entropy loss?

**Comment by WongKinYiu** [comment]
hmm... i check the code, it seems https://github.com/AlexeyAB/darknet/blob/master/src/yolo_layer.c#L259 should be modified.
